### PR TITLE
[FSDP2] Added forward unshard/wait for unshard/reshard

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118755
* #118223
* #118136
* #118118
* #118017
* #118004
* #117975
* __->__ #117973
* #117955
* #117950

This PR adds the all-gather and free logic required for forward.
- We define the logical all-gather as two ops: (1) unshard and (2) wait for unshard. This abstraction allows capturing both implicit forward prefetching (using multiple streams and `async_op=False`) and explicit forward prefetching (using `async_op=True`).
- Symmetrically, we define the reshard op to free the unsharded parameters.

Some other notes:
- The `FSDPParamGroup` and its `FSDPParam`s transition their sharded states together. This invariant allows us to reason about the parameters by group rather than individually with respect to whether they are sharded or unsharded.

---

### How Does the Overlap Work for All-Gather?

For context, the all-gather consists of three steps: (1) copy-in, (2) all-gather collective, and (3) copy-out.

<details>
<summary> Example </summary>

Suppose we have 2 parameters with shapes `(3, 3)` (denoted with `A`s) and `(2, 2)` (denoted with `B`s) and 2 ranks, where `P` represents padding and `E` represents empty:
```
Given:
(3, 3): AAAAAAAAA
(2, 2): BBBB

Sharded parameters/all-gather inputs:
Rank 0: AAAAAA, BB
Rank 1: AAAPPP, BB

Each rank allocate group's all-gather output:
EEEEEEEEEEEEEEEE
Each rank copy-in:
Rank 0: AAAAAABBEEEEEEEE
Rank 1: EEEEEEEEAAAPPPBB

Each rank all-gather:
Rank 0: AAAAAABBAAAPPPBB
Rank 1: AAAAAABBAAAPPPBB

Each rank copy-out:
Rank 0: AAAAAAAAAPPP, BBBB
Rank 1: AAAAAAAAAPPP, BBBB
```
</details>

`dist.all_gather_into_tensor()` always has the PG's NCCL stream wait for the current stream before running the collective. `async_op=True` means that the function waits on the work, having the current stream wait for the NCCL stream before returning. `async_op=False` means it returns the `Work` object, which the user can wait on later.

#### Implicit Prefetching
Implicit prefetching achieves communication/computation overlap without changing the CPU issue order:
- We use separate streams for copy-in and for issuing the `dist.all_gather_into_tensor()`. The copy-in stream allows us to overlap the copy-in with all-gather/reduce-scatter in backward, and the all-gather stream allows us to overlap the all-gather with forward compute (issued before it).
     - Because `dist.all_gather_into_tensor()` always has the PG's NCCL stream wait for the current stream, we need this "dummy" all-gather stream to prevent the all-gather from waiting on the forward compute with which it should overlap.
     - Without the separate copy-in stream, we cannot overlap all-gather copy-in with all-gather in forward.
- We copy-out in the default stream after having the default stream wait for the all-gather. This means that the autograd leaves are allocated in the default stream and autograd will not call `recordStream`.

Implicit prefetching does not require knowing the execution order ahead of time. However, when overlapping the next all-gather with the current compute, there may be a gap from the CPU thread issuing the current compute. If the CPU thread can run ahead, then this is not an issue.

#### Explicit Prefetching
Explicit prefetching achieves communication/computation by changing the CPU issue order, namely by reordering the all-gather to be before the compute with which it should overlap.
- Because we reorder, we do not need any separate streams, and we can use `async_op=False` for overlap.
- We can expose this explicit prefetching as a module-level `unshard()` op (e.g. `module.unshard(async_op: bool)`, and we can use it as a primitive for implementing the explicit forward prefetching in existing FSDP.

Explicit prefetching requires knowing the execution order.

---

Disclaimer: The testing is relatively lighter in this PR. I did not want to spend too much time writing new forward-only tests. The stream usage will be exercised thoroughly once we have backward too.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Review by weifengpy — COMMENTED** [review]
will approve after CI's confirming of no ``state_ref`` error

**Comment by awgu** [comment]
@weifengpy Rebasing is hard 😢 

I fixed the `state_ref` error.

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 1]
```diff
@@ -65,6 +74,25 @@ def _lazy_init(self) -> None:
...
+        self._init_shared_state()
+
+    def _init_shared_state(self) -> None:
+        # Setting the all-gather/reduce-scatter streams to be higher priority
+        # can help avoid some issues where their copies in/out are delayed and
+        # block computation
+        high_priority = -1
+        self._default_stream = torch.cuda.current_stream()
+        self._all_gather_copy_in_stream = torch.cuda.Stream(priority=high_priority)
+        self._all_gather_stream = torch.cuda.Stream(priority=high_priority)
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:86]: Curious about the purpose of `_all_gather_stream`. Since collectives kernels always run on nccl pg's internal stream, `_all_gather_stream` would just be an indirection for stream synchronization? Also IIUC the priority specified here won't affect the comm kernels.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:86]: > Also IIUC the priority specified here won't affect the comm kernels.

Correct! The priority here does not affect the comm. kernels.

> Curious about the purpose of _all_gather_stream. Since collectives kernels always run on nccl pg's internal stream, _all_gather_stream would just be an indirection for stream synchronization?

Also correct! The `_all_gather_stream` here is a trick to deal with the fact that NCCL PG collectives sync the NCCL stream with the current stream before running the collective.

This `_all_gather_stream` is used for so-called implicit prefetching where the CPU issue order is still the normal order: layer 0 all-gather -> layer 0 forward compute -> layer 1 all-gather -> layer 1 forward compute etc.

If we did not have a separate `_all_gather_stream`, then calling layer 1 all-gather would have the NCCL stream sync with the default (compute) stream and undesirably wait for layer 0 forward compute. The `_all_gather_stream` really just serves as a dummy stream so that we do not introduce that false dependency between layer i all-gather and layer i-1 forward compute.

Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:86]: Got it. This makes sense.

Would it make sense to fold `_all_gather_copy_in_stream` and `_all_gather_stream` into a single stream, so we use fewer streams and simpify synchronization? Say we have a single stream `s`:
- `s` waits for the param group to be updated by the optimizer
- Allocate all-gather input buffer on `s`
- Perform copy-in on `s`
- Issue and wait all-gather from `s`

Would this work/does this make sense?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:86]: Having both the AG copy-in stream and AG stream allows overlapping the _next_ all-gather copy-in with the _current_ all-gather, which can improve throughput in the forward pass. (In the backward pass, the all-gather copy-in can overlap with reduce-scatter.)

You might wonder, does overlapping the next all-gather copy-in with the current all-gather require using more memory? The answer is yes! However, I felt that it was a worthy tradeoff because:
- In the forward pass, including this overlap increases FSDP's peak memory usage from ~2x 'layer' size to 3x: current all-gather buffer (running all-gather), current copy-out parameters, next all-gather buffer (for copy-in).
- However, for the typical case without frozen parameters, the backward pass has FSDP peak memory of ~3x 'layer' size anyway! (e.g., prefetch next all-gather, current parameters, and current gradients)
- Even in the frozen parameter case, the ~1x 'layer' size is arguably in the noise of existing FSDP's `recordStream` spikes, and the real peak memory is generally not in forward. As such, I would lean toward just doing this overlapping always. If it is a really large concern, we could consider exposing it as an option to the user.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:86]: Using one combined all-gather stream:
Screenshot 2024-01-30 at 7 50 24 PM

Using separate all-gather copy-in and all-gather streams:
Screenshot 2024-01-30 at 7 51 35 PM

Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:86]: Thanks for the elaborate answer!

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -28,9 +36,141 @@ def __init__(
...
+        # Group's sharded state always matches its parameters' sharded states
         self._sharded_state = ShardedState.SHARDED
+        self._init_mp_dtypes()
         self._module_fqn: Optional[str] = None  # prefixed from root module
 
+        # - Communication and communication/computation overlap
+        default_stream = torch.cuda.current_stream()
+        self.default_stream: torch.cuda.Stream = default_stream
+        self.all_gather_copy_in_stream: torch.cuda.Stream = default_stream
+        self.all_gather_stream: torch.cuda.Stream = default_stream
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:48]: why we by default make these all default stream? can we just make them Nones and let root state shared state assign the correct streams to all param groups?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 3]
```diff
@@ -14,6 +14,25 @@ class AllGatherResult(NamedTuple):
     all_gather_input_numels: List[int]
 
 
+class AllGatherState(NamedTuple):
+    all_gather_result: AllGatherResult
+    event: torch.cuda.Event  # copy-out
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:19]: The `AllGatherState` is shared across `FSDPParamGroup`s to implement overlapping copy-in with all-gather. We need to keep the current all-gather result alive until we copy-in for the next all-gather.

Because this is some shared object across `FSDPParamGroup`s, I had it as a separate class. Alternatively, we could implement it as something like a singleton list, but I felt that that might be less clear.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 4]
```diff
@@ -28,9 +36,141 @@ def __init__(
...
+        else:
+            self._wait_all_gather_streams_on_event(all_gather_copy_out_event)
+        self._all_gather_result = None  # free unless saved in `all_gather_state`
+
+    def _wait_all_gather_streams_on_event(self, event: torch.cuda.Event):
+        self.all_gather_copy_in_stream.wait_event(event)
+        self.all_gather_stream.wait_event(event)
+
+    def reshard(self):
+        self._to_sharded()
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:118]: I think `reshard` is a higher-level abstraction. There will be more code going in here after we add `reshard_after_forward` arg support.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:48]: It is for type checking :)

Having something be `Optional[T]` is a pain for mypy.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:48]: Ah, sorry -- I also refactor this in a later PR: https://github.com/pytorch/pytorch/pull/118004

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 5]
```diff
@@ -14,6 +14,25 @@ class AllGatherResult(NamedTuple):
...
+    def put(self, state: AllGatherState) -> None:
+        assert self._state is None, "Expects to hold only one all-gather state"
+        self._state = state
+
+    def pop(self) -> Optional[AllGatherState]:
+        state = self._state
+        self._state = None
+        return state
+
+
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:35]: Is there an indirect synchronization that ensures that the first all-gather happens after the optimizer step for the corresponding parameters completes?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:35]: Sorry, since this PR does not cover forward/backward training, this synchronization step is missing in this PR.

It will be done in the first forward to run (almost always the root's forward).
https://github.com/pytorch/pytorch/pull/118004/files#diff-a2be7eb5a3d826e913056d5cf851e94be0914e8cca00dfb1a820fbae747eb494R73

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:19]: I deleted the original comment to try to re-explain.

This `AllGatherState` and `AllGatherStateHolder` are used to implement overlapping the next all-gather copy-in with the current all-gather. To do so, we must keep the current all-gather result alive until that copy-in finishes to avoid reusing the memory. This is why this `AllGatherState` has an `all_gather_result` field but never uses it. It is only for the reference.

The `AllGatherStateHolder` object needs to be shared across all `FSDPParamGroup`s as the next parameter group is responsible for freeing the current's all-gather result.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 6]
```diff
@@ -28,9 +36,141 @@ def __init__(
...
+
+    def post_forward(self, module: nn.Module, input: Any, output: Any):
+        with torch.profiler.record_function("FSDP::post_forward"):
+            self.reshard()
+            self._training_state = TrainingState.IDLE
+            return output
+
+    # Utilities #
+    def _to_sharded(self):
+        if self._sharded_state != ShardedState.SHARDED:
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:137]: Curious if it's valid to call `_to_sharded` in the `ShardedState.SHARDED` state. If not, shall we error out here?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 7]
```diff
@@ -14,6 +14,25 @@ class AllGatherResult(NamedTuple):
     all_gather_input_numels: List[int]
 
 
+class AllGatherState(NamedTuple):
+    all_gather_result: AllGatherResult
+    event: torch.cuda.Event  # copy-out
+
+
+class AllGatherStateHolder:
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:22]: We need some kind of shared container class. We could remove `put` and `pop`, but we would still need the class of some kind.

Using a `list` could work, but it is not obvious that the `list` should only have 0 or 1 elements.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 6 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:137]: It is valid and should be a no-op. For example, for state dict, we want the pre-save and pre-load hooks to call `_to_sharded` to make sure we are exposing the sharded `DTensor` parameters. We prefer to just have the hooks call `_to_sharded` unconditionally.

**Comment by awgu** [comment]
I realized that the next few PRs might prefer to be reviewed together. A lot of refactoring happens in https://github.com/pytorch/pytorch/pull/118004 because we have the whole communication infra setup then.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 8]
```diff
@@ -28,9 +36,141 @@ def __init__(
...
+        if len(orig_dtypes) != 1:
+            # This can be relaxed if we copy-out for the reduce-scatter
+            raise AssertionError(
+                f"FSDP expects uniform original parameter dtype but got {orig_dtypes}"
+            )
+        self._orig_dtype = next(iter(orig_dtypes))
+        self._param_dtype = self._orig_dtype
+
+    # Runtime #
+    def unshard(self, async_op: bool = False):
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:67]: Curious under what circumstance is `unshard` called multiple times within an iteration for one `TrainingState`. I'm trying to understand why we want to make the method idempotent instead of erroring out on re-entrant.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:67]: I think you are right that we should not call `unshard()` more than once for the same `TrainingState`. I can think on this more.

However, idempotence (across `unshard` calls with different `TrainingState`s) gives us one nice thing: we can have an `async_op=True` call before an `async_op=False` call.

Suppose the user wants to _manually_ prefetch the 1st all-gather that would otherwise not be overlapped with any compute. The user can add a `fsdp_module.unshard(async_op=True)` call, e.g. before some embedding all-to-all. Then, when the `fsdp_module` runs `forward()` per normal, it will call `unshard(async_op=False)`, return early since `self._all_gather_result is not None`, and call `wait_for_unshard()`, waiting on the 1st `async_op=True` call.

This `fsdp_module.unshard()` method has not been exposed from the `FSDP` class yet in these PRs, but I do plan to do that in the future. 

Let me know if this makes sense.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:19]: I am working on refactoring this in https://github.com/pytorch/pytorch/pull/118004.

**Review by yifuwang — APPROVED** [review]
Looks great!

I find the state transition really useful for reasoning about the logic! I think it would also be useful to document the invariants of each state. This way, future readers can understand the logic within the context of a single state without having to understand the whole picture.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 8 above)
Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:67]: Thanks for the context. It's totally fine to leave this idempotent so long as there's a useful scenario.

**Comment by shadow150519** [comment]
> > @awgu hi,  could you tell me why we do not need separate stream and can use async_op=False for overlap if we do explicit prefetch? Won't async_op=False call `work.wait()` and block default stream until comm is finished?
> 
> i think this is a typo. explicit prefetch and async_op=True are used together. all-gather happens in nccl steam, work.wait happens in the default stream

@weifengpy  thanks for your reply! I agree there maybe some typos in the original description. It also seems that explicit prefetch isn’t strictly tied to using a separate stream, since explicit prefetch merely triggers the unshard call in advance. Both of the following approaches can work:
+ Separate stream with async_op=False
+ Default stream with async_op=True

Could you share more about what advantages you see in using the default stream? I believe one possible benefit might be earlier memory reclamation (i.e., freeing unused GPU memory sooner), but I’d love to hear your perspective.

**Comment by weifengpy** [comment]
> Could you share more about what advantages you see in using the default stream? I believe one possible benefit might be earlier memory reclamation (i.e., freeing unused GPU memory sooner), but I’d love to hear your perspective.

when async_op=True, all_gather buffers are allocated in the default stream. this avoids inter-steam memory fragmentation. all tensors (compute and AG) are allocated in the default memory pool
